lib/web_ui/lib/src/engine/text/layout_service.dart

```diff
@@ -1115,81 +1001,55 @@ class LineBuilder {
 
-  late ui.TextDirection _currentBoxDirection = _paragraphDirection;
+  void addFragment(MeasuredFragment fragment) {
+    fragment = _updateMetrics(fragment);
 
-  late ui.TextDirection _currentContentDirection = _paragraphDirection;
+    if (fragment.isBreak) {
+      _lastBreakableFragment = _fragments.length;
+    }
 
-  bool _shouldCreateBoxBeforeExtendingTo(DirectionalPosition newEnd) {
-    // When the direction changes, we need to make sure to put them in separate
-    // boxes.
-    return newEnd.isSpaceOnly || _currentBoxDirection != newEnd.textDirection || _currentContentDirection != newEnd.textDirection;
+    _fragments.add(fragment);
   }
 
-  /// Extends the line by setting a [newEnd].
-  void extendTo(DirectionalPosition newEnd) {
-    ascent = math.max(ascent, spanometer.ascent);
-    descent = math.max(descent, spanometer.descent);
-
-    // When the direction changes, we need to make sure to put them in separate
-    // boxes.
-    if (_shouldCreateBoxBeforeExtendingTo(newEnd)) {
-      createBox();
+  /// Updates the [LineBuilder]'s metrics to take into account the new [fragment].
+  MeasuredFragment _updateMetrics(MeasuredFragment fragment) {
+    if (fragment.isSpaceOnly) {
+      _spaceBoxCount++;
+    } else {
+      width = widthIncludingSpace + fragment.metrics.widthExcludingTrailingSpaces;
     }
-    _currentBoxDirection = newEnd.textDirection ?? _currentBoxDirection;
-    _currentContentDirection = newEnd.textDirection ?? ui.TextDirection.ltr;
+    widthIncludingSpace += fragment.metrics.widthIncludingTrailingSpaces;
 
-    _addSegment(_createSegment(newEnd.lineBreak));
-    if (newEnd.isSpaceOnly) {
-      // Whitespace sequences go in their own boxes.
-      createBox(isSpaceOnly: true);
+    if (fragment.isPlaceholder) {
+      return _updateHeightForPlaceholder(fragment);
     }
-  }
 
-  /// Extends the line to the end of the paragraph.
-  void extendToEndOfText() {
-    if (end.type == LineBreakType.endOfText) {
-      return;
+    if (fragment.isBreak) {
+      _breakCount++;
     }
 
-    final LineBreakResult endOfText = LineBreakResult.sameIndex(
-      paragraph.toPlainText().length,
-      LineBreakType.endOfText,
-    );
-
-    // The spanometer may not be ready in some cases. E.g. when the paragraph
-    // is made up of only placeholders and no text.
-    if (spanometer.isReady) {
-      ascent = math.max(ascent, spanometer.ascent);
-      descent = math.max(descent, spanometer.descent);
-      _addSegment(_createSegment(endOfText));
-    } else {
-      end = endOfText;
-    }
+    ascent = math.max(ascent, fragment.metrics.ascent);
+    descent = math.max(descent, fragment.metrics.descent);
+    return fragment;
   }
 
-  void addPlaceholder(PlaceholderSpan placeholder) {
-    // Increase the line's height to fit the placeholder, if necessary.
+  MeasuredFragment _updateHeightForPlaceholder(MeasuredFragment fragment) {
+    final PlaceholderSpan placeholder = fragment.span as PlaceholderSpan;
```

Comment:
Similarly here. It feels like we should know ahead of time that the span is a `PlaceholderSpan` without a cast 🤔

Reply:
This one I couldn't figure out a clean way to remove the cast. If you have ideas, I'm happy to consider.